sqlite-db/src/rollover/overwrite.rs

```diff
@@ -16,64 +15,47 @@ use sqlx::Connection as SqlxConnection;
 use sqlx::Transaction;
 
-pub async fn insert(
+mod delete;
+
+/// Overwrite a CFD's latest rollover data.
+///
+/// After a successful rollover, we can forget about the previous `Dlc`, `FundingFee` and
+/// `CompleteFee`.
+pub async fn overwrite(
     connection: &mut PoolConnection<Sqlite>,
     event_id: i64,
-    event: CfdEvent,
+    order_id: models::OrderId,
+    dlc: Dlc,
+    funding_fee: FundingFee,
+    complete_fee: Option<CompleteFee>,
 ) -> Result<()> {
-    let event_kind = event.event;
-    match event_kind {
-        EventKind::RolloverCompleted {
-            dlc: Some(dlc),
-            funding_fee,
-            complete_fee,
-        } => {
-            let mut inner_transaction = connection.begin().await?;
-
-            crate::rollover::delete::delete(&mut inner_transaction, event.id.into()).await?;
+    let mut inner_transaction = connection.begin().await?;
 
-            insert_rollover_completed_event_data(
-                &mut inner_transaction,
-                event_id,
-                &dlc,
-                funding_fee,
-                complete_fee,
-                event.id.into(),
-            )
-            .await?;
+    delete(&mut inner_transaction, order_id).await?;
 
-            for revoked in dlc.revoked_commit {
-                insert_revoked_commit_transaction(&mut inner_transaction, event.id.into(), revoked)
-                    .await?;
-            }
+    insert_rollover_completed_event_data(
+        &mut inner_transaction,
+        event_id,
+        &dlc,
+        funding_fee,
+        complete_fee,
+        order_id,
+    )
+    .await?;
 
-            for (event_id, cets) in dlc.cets {
-                for cet in cets {
-                    insert_cet(
-                        &mut inner_transaction,
-                        event_id.into(),
-                        event.id.into(),
-                        cet,
-                    )
-                    .await?;
-                }
-            }
+    for revoked in dlc.revoked_commit {
+        insert_revoked_commit_transaction(&mut inner_transaction, order_id, revoked).await?;
+    }
 
-            // Commit the transaction to either write all or rollback
-            inner_transaction.commit().await?;
-        }
-        EventKind::RolloverCompleted { dlc: None, .. } => {
-            // We ignore rollover completed events without DLC data as we don't need to store
-            // anything
-            tracing::error!(
-                "Invalid RolloverCompleted event: Trying to insert a RolloverCompleted event without a DLC"
-            )
-        }
-        _ => {
-            tracing::error!("Invalid event type. Use `append_event` function instead")
+    for (event_id, cets) in dlc.cets {
+        for cet in cets {
+            insert_cet(&mut inner_transaction, event_id.into(), order_id, cet).await?;
         }
     }
 
+    // Commit the transaction to either write all or rollback
+    inner_transaction.commit().await?;
```

Comment:
@luckysori you mention here that it should either commit or rollback, however I don't see an error handling here that would rollback. Is this somewhere hidden in the rust magic? furthermore, do we need a bit more error handling here - otherwise a failed transaction will remain failed. Do we store the events - so we can keep a status or something?

Reply:
I didn't write this comment, I just moved it around 😛 Tbh, I think the comment is a bit confusing/unnecessary.

> you mention here that it should either commit or rollback, however I don't see an error handling here that would rollback. Is this somewhere hidden in the rust magic?

The library we are using [handles the rollback for us](https://docs.rs/sqlx/latest/sqlx/struct.Transaction.html).

> furthermore, do we need a bit more error handling here - otherwise a failed transaction will remain failed. Do we store the events - so we can keep a status or something?

If we fail, we roll back and bubble up an error. Since the `RolloverCompleted` event doesn't get inserted, the state of the CFD won't move forward for this party. Assuming that it did succeed for the other party, they would eventually need to catch up via the "rollover retry" feature that we introduced recently.

Basically, failing to overwrite this state is not handled very well. I've opened https://github.com/itchysats/itchysats/issues/2662 to talk about it.